Lifecycle of a transaction doc (#1)
glozow/bitcoin-notes: transaction-lifecycle.md

[freezes chainstate](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L1022) for the duration of this validation session to ensure those values don't change.
Even if a new block (which could extend or invalidate the chain tip) arrives on the wire, the
node won't process it until it has finished validating this transaction.

#### Signature and Script Checks

Relevant code: [src/script/interpreter.h](https://doxygen.bitcoincore.org/interpreter_8h.html), [`CheckInputScripts()`](https://doxygen.bitcoincore.org/validation_8cpp.html#a6a96a3e1e6818904fdd5f51553b7ea60), [`PrecomputedTransactionData`](https://doxygen.bitcoincore.org/struct_precomputed_transaction_data.html),
[`g_scriptExecutionCache`](https://doxygen.bitcoincore.org/validation_8cpp.html#ab821bded1c098da86bb13a4ac9d1208b), [`CScriptCheck`](https://doxygen.bitcoincore.org/class_c_script_check.html), [`STANDARD_SCRIPT_VERIFY_FLAGS`](https://doxygen.bitcoincore.org/policy_8h.html#ab28027bf27efcdd6535a13175a89ca5a)

> Transaction [script checks](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L1356) are actually context-free; the unlocking script ([`scriptSig`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L1357),

Author: Will never use these words again in my life :salute:

[serialized, hashed, and stored](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/script/interpreter.cpp#L1423) in a [`PrecomputedTransactionData`](https://doxygen.bitcoincore.org/struct_precomputed_transaction_data.html) struct for use in signature
verification.

[`MemPoolAccept`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L426) performs two sets of script checks: [`PolicyScriptChecks`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L917) and
[`ConsensusScriptChecks`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L943). The former [runs the script interpreter](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L926) using [consensus and policy flags](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L922) and

> [caches the result](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/script/sigcache.cpp#L114) in the [signature cache](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/script/sigcache.cpp#L90). The latter runs the script interpreter using [consensus

Author: AFAIK the script check result is cached in `ConsensusScriptChecks`?

> [caches the result](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/script/sigcache.cpp#L114) in the [signature cache](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/script/sigcache.cpp#L90). The latter runs the script interpreter using [consensus

Author: Oh nvm I see what you mean

   have), also identified by their hash.

3. The sender [sends the full transaction](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L1879) in a `TX` message.

### Transaction Announcement and Broadcast

Relevant code: [`CTxMemPool::m_unbroadcast_txids`](https://doxygen.bitcoincore.org/class_c_tx_mem_pool.html#a3df5ff43adfe0f407daa6fdef8965ba8), [`PeerManagerImpl::ReattemptInitialBroadcast()`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L1099)

With respect to privacy in transaction relay, we specifically want to hide the origin (IP address)
of a transaction. For each peer, nodes send a batch of transaction announcements at [random intervals](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L4584-L4588)
(poisson timer with 2-second average for outbound peers and 5-second average for inbound peers).

After broadcasting a transaction for the first time, Bitcoin Core [nodes store them](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/node/transaction.cpp#L101) in an

> _unbroadcast_ set until they see an `INV` for it, indicating that the transaction has been accepted

Reviewer: The transaction is removed when the node receives a `getdata` for it, not an `inv`.

Author: Gah oops

(ancestor feerate + manual prioritization by the miner), to determine which transactions to [include
in the next block](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/miner.cpp#L226). This means a "package" consisting of a 5 sat/vB parent and 100 sat/vB child may
get picked over a 20 sat/vB transaction, but not over a singular 100 sat/vB transaction with no
dependencies.

### Block Relay

Relevant code: [BIP152](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki)

Blocks can contain megabytes worth of transactions, and nodes that keep a mempool have seen
at least 98% of them already (FIXME source: [gmax's block relay talk](https://www.youtube.com/watch?v=EHIuuKCm53o)).

The methods for block relay are listed below, along with a description of how the transaction is
communicated with or without sending the full transaction data over P2P. Note that peers negotiate

> which block relay communication method to use _before_ blocks are communicated.

Reviewer: ```suggestion
which block relay communication method when initially establishing the connection.
```

- High Bandwidth Compact Blocks: the sender [creates a BIP152 compact block](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L1481) comprising only the block
  header, prefilled coinbase transaction, and shortids for all other transactions. If the
receiver does not recognize a transaction shortid, it [may request](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L3434) it through `GETBLOCKTXN`.

- Headers First: to prevent huge spikes in network traffic, senders flood [block headers only](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L4526) (which
  is just 80 bytes). The recipient can choose to request the [full block](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L2093) or [compact block](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L2107).

> - `INV`/`GETDATA` Dialogue: the peers sacrifice latency for lower bandwidth usage by announcing blocks [through

Reviewer: In Bitcoin Core, we only revert to announcing via `inv` if the blocks that we're trying to announce don't connect to each other or we don't think they connect to the peer's block chain (or if the peer does not support headers-first relay). I don't think latency/bandwidth tradeoff is the reason here.

>   `INV` messages](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/net_processing.cpp#L1813). The recipient can request the full or compact block.

Reviewer: In Bitcoin Core, we'll never respond to a block `inv` with a request for the block or compact block. We'll ask for the headers connecting to that blockhash: https://github.com/bitcoin/bitcoin/blob/842e2a9c54bd27865712b154a1582f6f368713f1/src/net_processing.cpp#L2883

### Block Validation

Relevant code: [`-par`](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/init.cpp#L395), [`g_scriptExecutionCache`](https://doxygen.bitcoincore.org/validation_8cpp.html#ab821bded1c098da86bb13a4ac9d1208b), [`CChainState::ActivateBestChain()`](https://doxygen.bitcoincore.org/class_c_chain_state.html#af771ed84553b1fb7269cfdfb54f66c48)

FIXME: non-script checks

If the node [already validated this transaction before](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L1490) it was included in a block, no consensus rules
have changed, and the script cache has not evicted this transaction's entry, it doesn't need to [run
script checks again](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/validation.cpp#L1428-L1430)!

> ### State Changes and Persistence to Disk

Reviewer: This section or the one above could explicitly talk about updating the utxo set with the effects of the transaction (remove the UTXOs it spends and add the outputs it creates)

implementing the node's [`ValidationInterface`](https://doxygen.bitcoincore.org/class_c_validation_interface.html). The most relevant information for the wallet is which
coins are available and how likely they are to remain spendable - this is quantified by the number
of confirmations on the transaction (or on a [_conflicting_](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/transaction.h#L342) transaction, [represented as negative
numbers](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/wallet.cpp#L1128)). A UTXO made available by a transaction 100 blocks deep in the most-work chain is
considered pretty safe. On the other hand, if a newly confirmed transaction conflicts with one that
sent coins to the wallet, those coins now have a "depth" of -1 and [will not be considered a probable
source of funds](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/spend.cpp#L94).

> ### Transaction Children

Reviewer: I'm not sure this should be a different section at the same level as the other sub-headings.

A transaction has a "child" when its outputs are used as the inputs to fund another transaction
(which may or may not be valid). Technically, a child can be created as soon as the transaction has

> a txid (i.e. after the transaction has been signed), including some limitations:

Reviewer: A transaction can have a txid before it's been signed (e.g. a segwit transaction without the witness attached yet).

Author: Oops :facepalm: I probably meant to say something like "this can happen before the tx has been signed" and wrote the opposite

permissions or unreasonable fees. However, the permissionless nature of the P2P network also means
that nodes are exposing their computational resources to peers that may try to abuse them.
Malicious nodes can create fake transactions very cheaply (both monetarily and computationally);
there are no Proof of Work requirements on transactions.

### Mempool Validation

Selecting the best transactions for the resource-constrained mempool involves a tradeoff between
applying consensus rules correctly and protecting the node from DoS attacks. As such, apply a set of
validation rules known as mempool _policy_ (also sometimes called "standardness" or "non-mandatory"
rules) in addition to consensus.

We might categorize transaction validation checks in a few different ways:

- Consensus vs Policy: These can also be thought of as mandatory vs non-mandatory checks. These two
  are not mutually exclusive, but we make all possible efforts to compartamentalize consensus rules
to avoid making mempool logic consensus-critical.

- Script vs Non-script: [Script](https://en.bitcoin.it/wiki/Script) refers to the data in scriptSig,

>   scriptPubKey, and witness that specifies spending conditions. We make this distinction because

Reviewer 2: The witness is not a script but it can contain an element that is a script. For P2WSH the last witness element is the witness script and in P2TR script path spends the penultimate element would be the leafscript.

in the same directory. If the node operator has configured a maximum memory allocation for this
node, this process will also prune blocks by oldest-first to stay within limits.

While the UTXO set is comparatively smaller than the block chain, it doesn't always fit in memory.
A node's [view of available coins]() is implemented in layers of maps from outpoints to
scriptPubKeys. Every validation session (for both [block
validation](https://github.com/bitcoin/bitcoin/blob/6312b8370c5d3d9fb12eab992e3a32176d68f006/src/validation.cpp#L2380)
on top of the current tip and unconfirmed [transaction
validation](https://github.com/bitcoin/bitcoin/blob/6312b8370c5d3d9fb12eab992e3a32176d68f006/src/validation.cpp#L427)
on top of current chainstate and mempool) creates a temporary view of the state and flushes
it to the appropriate view.

### Wallet Updates

The Bitcoin Core wallet tracks incoming and outgoing transactions and subscribes to major events by
[implementing](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/wallet.cpp#L1187-L1265)
the node's `ValidationInterface`. The most relevant information for the wallet is which of its own

> coins are available and how likely they are to remain spendable - this is quantified by the

Reviewer: Maybe add a sentence or two on how the wallet considers which coins are its own (either pattern-matching or using a descriptor).

[number](https://github.com/bitcoin/bitcoin/blob/55a156fca08713b020aafef91f40df8ce4bc3cae/src/wallet/spend.h#L25)
of confirmations on the transaction (or on a conflicting transaction, represented using a negative

> number). A UTXO made available by a transaction 100 blocks deep in the most-work chain is considered

Reviewer: Remove "or on a conflicting transaction..." - too low-level implementation detail

pretty safe. On the other hand, if a newly confirmed transaction conflicts with one that sent coins
to the wallet, those coins have a "depth" of -1 and [are not
considered](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/spend.cpp#L94)
a probable source of funds.

At the end of a transaction's lifecycle, it has deleted and added UTXOs to the network state, added
an entry to the blockchain, and faciliated a transfer of value between two Bitcoin users anywhere in
the world. While some nodes may prune the transaction data from their databases, its outputs can be

> used to fund other transactions in the future.

Reviewer: I don't think you should finish the article with a sentence about pruning. That seems quite a tangential detail to the main focus.

Maybe also add a summary section that ties everything together.

transactions to their Bitcoin Core node via the
[`sendrawtransaction`](https://developer.bitcoin.org/reference/rpc/sendrawtransaction.html) RPC.

### Transaction Creation through Bitcoin Core Wallet

The Bitcoin Core wallet also allows users to create transactions with varying levels of
customization. The steps are roughly as follows:

1. A recipient provides an [invoice or destination](https://en.bitcoin.it/wiki/Invoice_address) to
   pay to, committing to the spending conditions. This helps the wallet create
[outputs](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/outputtype.h#L23-L26),
which comprise the payments.

2. The wallet calculates feerate estimates,
output types, and other options based on the user's input, pre-configured defaults and preferences,
historical blocks and current mempool contents queried from the node.

3. The transaction is
    ["funded"](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/spend.cpp#L373)
by selecting inputs from the set of
[UTXOs](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/spend.cpp#L67)

> available in the wallet; these comprise the inputs. A change output [may or may not be

Author: good point
probably a bit too low-level to say one or more, so I'm gonna say "change may or may not be added" :sweat_smile:

containing a valid Proof of Work solution and accepted by the network as part of the most-work chain
- it is prohibitively expensive for a private key owner to reverse a transaction by rewriting (i.e.
recomputing Proofs of Work for) these blocks.

### Mining

Miners can start working on a block as soon as they have the previous block hash. The process of
creating a new block might look something like this:

1. The miner calls the Bitcoin Core RPC
   [`getblocktemplate`](https://developer.bitcoin.org/reference/rpc/getblocktemplate.html) to
determine the best set of transactions from the mempool that fit into a block's [weight and sigop
limits](https://github.com/bitcoin/bitcoin/blob/7fcf53f7b4524572d1d0c9a5fdc388e87eb02416/src/consensus/consensus.h#L14-L17).
This generates a [_block template_](https://github.com/bitcoin/bips/blob/master/bip-0022.mediawiki)
containing a consensus-valid set of transactions and block header, just mising the nonce.

2. The miner uses the block template to dispatch work tasks to other hardware (a dedicated rack of
   ASICs, a cloud instance, or other nodes op,erated by mining pool members) dedicated to finding the

> nonce. "Mining" typically refers to this task of guessing the nonce, specifically.

Author: good point

they are recalculated for each block prior to validation. Regardless of whether or not transactions
have already been previously validated and accepted to the mempool, nodes check block-wide consensus
rules (e.g. [total sigop
cost](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L1935),
[duplicate
transactions](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L1778-L1866),
[timestamps](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L3172-L3179),
[witness
commitments](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L3229-L3255),
[block subsidy
amount](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L1965-L1969))
and transaction-wide consensus rules (e.g. availability of inputs, locktimes, and [input
scripts](https://github.com/bitcoin/bitcoin/blob/9df1906091f84d9a3a2e953a0424a88e0931ea33/src/validation.cpp#L1946))
for each block.

Script checking is parallelized in block validation. Block transactions are checked in order (and

> coins set updated which allows for dependencies within the block), but input script

Reviewer 2: ```suggestion
coins get updated which allows for dependencies within the block), but input script
```

the node's
[`ValidationInterface`](https://github.com/bitcoin/bitcoin/blob/3d9cdb16897bf5f5eed525fd3fbc07e57dbe5f54/src/validationinterface.h#L63-L177).
The most relevant information for the wallet is which of [its own
coins](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/wallet.cpp#L1296-L1330)
are available and how likely they are to remain spendable, measured by the
[number](https://github.com/bitcoin/bitcoin/blob/55a156fca08713b020aafef91f40df8ce4bc3cae/src/wallet/spend.h#L25)
of confirmations on the transaction and its sender. A UTXO made available by a
transaction sent by the wallet itself, 100 blocks deep in the most-work chain, is considered pretty
safe. On the other hand, if a newly confirmed transaction conflicts with one that sent coins to the
wallet, those coins have a "depth" of -1 and [are not
considered](https://github.com/bitcoin/bitcoin/blob/1a369f006fd0bec373b95001ed84b480e852f191/src/wallet/spend.cpp#L94)
a probable source of funds.

### Conclusion

At the end of a transaction's lifecycle, it has deleted and added UTXOs to the network state, added

> an entry to the blockchain, and faciliated a transfer of value between two Bitcoin users anywhere in

Reviewer 2: ```suggestion
an entry to the blockchain, and facilitated a transfer of value between two or more Bitcoin users anywhere in
```
